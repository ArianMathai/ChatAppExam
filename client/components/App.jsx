import AppRoutes from "./AppRoutes";
import { LoginContext } from "../context/LoginContext";
import { useContext, useEffect, useState } from "react";
import { Link } from "react-router-dom";
import Logout from "./auth/Logout";

function App() {
  const [user, setUser] = useState();
  const [username, setUsername] = useState();
  const [email, setEmail] = useState();
  const [google_client_id, setGoogle_client_id] = useState();
  const [googleDiscoveryUrl, setGoogleDiscoveryUrl] = useState();
  const [microsoftEndpoint, setMicrosoftEndpoint] = useState();
  const [microsoftClientId, setMicrosoftClientId] = useState();

  async function loadUser() {
    const res = await fetch("/api/login/user");
    const usr = await res.json();
    setUser(usr);
    setUsername(usr.name);
    setEmail(usr.email);

    if (!user) {
      const response = await fetch("/api/login/microsoft/user");
      const micUser = await response.json();
      setUser(micUser);
      setUsername(micUser.name);
      setEmail(micUser.email);
    }
  }

  async function fetchLoginConfig() {
    const res = await fetch("/api/login/auth/config", {});

    const data = await res.json();

    setGoogle_client_id(data.client_id);
    setGoogleDiscoveryUrl(data.discoveryUrl);
    setMicrosoftClientId(data.microsoft_client_id);
    setMicrosoftEndpoint(data.microsoft_endpoint);
  }

  useEffect(() => {
    fetchLoginConfig();
  }, []);

  useEffect(() => {
    if (username === undefined) {
      loadUser();
    }
  }, [username]);

  return (
    <LoginContext.Provider
      value={{
        user,
        username: username,
        email: email,
        loadUser,
        google_client_id: google_client_id,
        google_discovery_url: googleDiscoveryUrl,
        microsoft_endpoint: microsoftEndpoint,
        microsoft_client_id: microsoftClientId,
      }}
    >
      <header>{username ? <h2>{username}</h2> : <h2> </h2>}</header>
      <nav>
        <Link to={"/"}>Home</Link>
        <Link to={"/chatroom"}>Chat</Link>
        <Link to={"/profile"}>Profile</Link>
        <Link to={"/profile/browse"}>Browse profiles</Link>
        <div style={{ flex: 1 }}></div>

        {!username ? (
          <Link to={"/login"}>Login</Link>
        ) : (
          <Logout className="logout-link" />
        )}
      </nav>

      <main>{microsoftEndpoint !== undefined && <AppRoutes />}</main>

      <footer>Made by Arian</footer>
    </LoginContext.Provider>
  );
}
export default App;
