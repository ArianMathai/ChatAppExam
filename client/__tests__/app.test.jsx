import React from 'react';
import renderer, {act} from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import App from '../components/App';

global.fetch = jest.fn(() =>
    Promise.resolve({
        ok: true,
        json: () => Promise.resolve({}),
    })
);

jest.mock('../components/auth/Logout.jsx', () => 'Logout');

describe('App Component', () => {
    it('should match the snapshot', async () => {
        let tree;
        await act(() => {
            tree = renderer.create(
                <MemoryRouter>
                    <App />
                </MemoryRouter>
            );
        });
        expect(tree.toJSON()).toMatchSnapshot();
    });
});
