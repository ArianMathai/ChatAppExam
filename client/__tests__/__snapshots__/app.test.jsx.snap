// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should match the snapshot 1`] = `
[
  <header>
    <h2>
       
    </h2>
  </header>,
  <nav>
    <a
      href="/"
      onClick={[Function]}
    >
      Home
    </a>
    <a
      href="/chatroom"
      onClick={[Function]}
    >
      Chat
    </a>
    <a
      href="/profile"
      onClick={[Function]}
    >
      Profile
    </a>
    <a
      href="/profile/browse"
      onClick={[Function]}
    >
      Browse profiles
    </a>
    <div
      style={
        {
          "flex": 1,
        }
      }
    />
    <a
      href="/login"
      onClick={[Function]}
    >
      Login
    </a>
  </nav>,
  <main />,
  <footer>
    Made by Arian
  </footer>,
]
`;
